src/runtime/virtcontainers/qemu.go

```diff
@@ -1570,7 +1571,7 @@ func (q *qemu) hotplugVFIODevice(ctx context.Context, device *config.VFIODev, op
 			case config.VFIODeviceNormalType:
 				err = q.qmpMonitorCh.qmp.ExecuteVFIODeviceAdd(q.qmpMonitorCh.ctx, devID, device.BDF, device.Bus, romFile)
 			case config.VFIODeviceMediatedType:
```

Comment:
This is the main consumer of the type value.  So it seems like it might make more sense to make AP device a separate option at that level, rather than testing for mediated type *then* for `isAP`.

Reply:
You make some valid concerns about API cleanliness, but there is a catch I'd like to hear your opinion on. 
@dgibson, you suggest breaking this down by `config.VFIODev.Type`, which, until now, can be a `VFIODeviceNormalType` or a `VFIODeviceMediatedType`, i.e. we'd have e.g. a `VFIOAPDeviceMediatedType` (is always mediated) (and perhaps rename the others).
Meanwhile, IIUC, both you and @jodh-intel suggest breaking down `config.VFIODev` (which also assumes PCI through and through) into two devices. (James, you said `enum` -- I guess you meant an `interface`?)
However, implementing both of these would introduce redundancy and possible illegal states (e.g. a `config.VFIOAPDev` whose `Type` is `VfioPCIDeviceNormalType` or such).
And we cannot break up the VFIO devices "from the beginning" (i.e. `DeviceManager.NewDevice`), because only after checking IOMMU groups can you know whether a device is PCI or AP. 
WDYT?

(Note that this is not the distinction of `kataDevice.Type`, which currently has `vfio` or `vfio-gk` (where `gk` is guest kernel, i.e. not for the container, which changes the agent-side handling, which I shouldn't think we'd need for AP). I guess adding an option there is not a huge conflict though.)

Reply:
The Go code, and the agent-protocol is already full of places which have representable invalid states.  One more doesn't make much difference.

I doubt you need a `vfio-gk` equivalent for AP.  `vfio` passes through devices in the way that a non-Kata container would do so: a VFIO device in the host becomes a VFIO device in the guest.  `vfio-gk` is a effectively poorly-defined Kata extension, which has the VFIO device in the host become.. something.. in the guest depending what driver is available in the guest kernel.